src/main/java/de/tum/in/www1/artemis/service/TutorialGroupService.java

```diff
@@ -0,0 +1,95 @@
+        this.courseRepository = courseRepository;
+        this.userService = userService;
+    }
+
+    /**
+     * Deregister a student from a tutorial group.
+     *
+     * @param student       The student to deregister.
+     * @param tutorialGroup The tutorial group to deregister from.
+     */
+    public void deregisterStudent(User student, TutorialGroup tutorialGroup) {
+        Optional<TutorialGroupRegistration> existingRegistration = tutorialGroupRegistrationRepository.findTutorialGroupRegistrationByTutorialGroupAndStudent(tutorialGroup,
+                student);
+        if (existingRegistration.isEmpty()) {
+            return; // No registration found, nothing to do.
+        }
+        tutorialGroupRegistrationRepository.delete(existingRegistration.get());
+    }
+
+    /**
+     * Register a student to a tutorial group.
+     *
+     * @param student       The student to register.
+     * @param tutorialGroup The tutorial group to register to.
+     */
+    public void registerStudent(User student, TutorialGroup tutorialGroup) {
+        Optional<TutorialGroupRegistration> existingRegistration = tutorialGroupRegistrationRepository.findTutorialGroupRegistrationByTutorialGroupAndStudent(tutorialGroup,
+                student);
+        if (existingRegistration.isPresent()) {
+            return; // Registration already exists, nothing to do.
+        }
+        TutorialGroupRegistration newRegistration = new TutorialGroupRegistration(student, tutorialGroup, TutorialGroupRegistrationType.INSTRUCTOR_REGISTRATION);
+        tutorialGroupRegistrationRepository.save(newRegistration);
+    }
+
+    /**
+     * Register multiple students to a tutorial group.
+     *
+     * @param courseId        The id of the course the tutorial group belongs to.
+     * @param tutorialGroupId The id of the tutorial group to register to.
+     * @param studentDTOs     The students to register.
+     * @return The students that could not be found and thus not registered.
+     */
+    public List<StudentDTO> registerMultipleStudents(Long courseId, Long tutorialGroupId, List<StudentDTO> studentDTOs) {
+        var course = this.courseRepository.findByIdElseThrow(courseId);
+        var tutorialGroup = tutorialGroupRepository.findByIdWithTeachingAssistantAndRegistrationsElseThrow(tutorialGroupId);
+
+        List<StudentDTO> notFoundStudentsDTOs = new ArrayList<>();
+        for (var studentDto : studentDTOs) {
+            var registrationNumber = studentDto.getRegistrationNumber();
+            var login = studentDto.getLogin();
+            Optional<User> optionalStudent = userService.findUserAndAddToCourse(registrationNumber, course.getStudentGroupName(), Role.STUDENT, login);
+            if (optionalStudent.isEmpty()) {
+                notFoundStudentsDTOs.add(studentDto);
+            }
+            else {
+                registerStudent(optionalStudent.get(), tutorialGroup);
```

Comment:
Registration happens very early just in the beginning of the semester, and rarely throughout the course. I will keep this solution therefore.

Reply:
I agree with @Hialus

Reply:
Just because it happens rarely doesn't mean you have to write it purposefully inefficiently. Unless the change is too complex, we should always take the more performant approach (especially when it comes to the DB).
In this case this would only be about extracting part of the registerStudent method into a helper method and collecting them here. This should only take about 5 minutes and decreases the amount of DB calls by around a factor of 20, so you should definitely do this.

Reply:
Done in 1f00c8c9e9259eacd45b6c44bbb9590de5186fde